### PR TITLE
Add shuffle for pytorch Dataloader

### LINKED ISSUE #341: Add shuffle for pytorch dataloader
Since the shuffling is not well supported by petastorm pytorch dataloader.  We need to improve the shuffle feature.

### PR DESCRIPTION
**Pull request by DeliangFan**
This fix #341

### CODE REVIEW

**Comment by DeliangFan** [comment]
@selitvin Would you please take a review

**Comment by selitvin** [comment]
Thank you for your PR! Would you be able to cover the new feature with a test?

**Comment by DeliangFan** [comment]
> Thank you for your PR! Would you be able to cover the new feature with a test?

Hi, @selitvin , I have pushed a new PR with some unit tests, please take a review.

File: petastorm/tests/test_pytorch_dataloader.py [hunk 1]
```diff
@@ -72,6 +72,22 @@ def test_simple_read(synthetic_dataset, reader_factory):
         _check_simple_reader(loader, synthetic_dataset.data, BATCHABLE_FIELDS - {TestSchema.sensor_name})
 
 
+@pytest.mark.parametrize('reader_factory', ALL_READER_FLAVOR_FACTORIES)
+def test_simple_read_with_small_shuffle(synthetic_dataset, reader_factory):
+    with DataLoader(reader_factory(synthetic_dataset.url, schema_fields=BATCHABLE_FIELDS,
+                                   transform_spec=TransformSpec(_sensor_name_to_int)),
+                    shuffle_size=20) as loader:
```
Comment (selitvin) [inline petastorm/tests/test_pytorch_dataloader.py]: Can you please verify that the shuffling is actually happening?
This test implementation may help: https://github.com/uber/petastorm/blob/a7234bc8af30762676129099811643f9451f938a/petastorm/tests/test_end_to_end.py#L248

File: petastorm/pytorch.py [hunk 2]
```diff
@@ -97,7 +98,7 @@ class DataLoader(object):
     runs out of samples.
     """
 
-    def __init__(self, reader, batch_size=1, collate_fn=decimal_friendly_collate):
+    def __init__(self, reader, batch_size=1, collate_fn=decimal_friendly_collate, shuffle_size=0):
```
[the remark this answers is not in the document]
Comment (DeliangFan) [reply petastorm/pytorch.py]: done~

File: petastorm/pytorch.py [hunk 3]
```diff
@@ -108,25 +109,33 @@ def __init__(self, reader, batch_size=1, collate_fn=decimal_friendly_collate):
...
-                batch = []
-        if batch:
-            yield self.collate_fn(batch)
+            self.batch.append(row_as_dict)
+
+            if len(self.batch) < self.shuffle_size:
+                continue
+
+            if len(self.batch) >= self.batch_size:
+                random.shuffle(self.batch)
```
[the remark this answers is not in the document]
Comment (DeliangFan) [reply petastorm/pytorch.py]: That sounds great, I will try to reuse it.

File: petastorm/tests/test_pytorch_dataloader.py (hunk 1 above)
Comment (DeliangFan) [reply petastorm/tests/test_pytorch_dataloader.py]: Done, Thanks

File: petastorm/pytorch.py [hunk 4]
```diff
@@ -108,25 +109,33 @@ def __init__(self, reader, batch_size=1, collate_fn=decimal_friendly_collate):
...
+                yield self.collate_fn(self.batch[0:self.batch_size])
+                self.batch = self.batch[self.batch_size:]
+
+        if self.batch:
```
[the remark this answers is not in the document]
Comment (DeliangFan) [reply petastorm/pytorch.py]: Done

**Comment by DeliangFan** [comment]
Hi, @selitvin  I have made some changes according your comments, would you please take a review again. Thanks a lot!

File: petastorm/tests/test_pytorch_dataloader.py [hunk 5]
```diff
@@ -72,6 +84,59 @@ def test_simple_read(synthetic_dataset, reader_factory):
...
 
+@pytest.mark.parametrize('reader_factory', ALL_READER_FLAVOR_FACTORIES)
+def test_simple_read_with_small_shuffle_size(synthetic_dataset, reader_factory):
+    with DataLoader(reader_factory(synthetic_dataset.url, schema_fields=BATCHABLE_FIELDS,
+                                   transform_spec=TransformSpec(_sensor_name_to_int)),
+                    shuffle_size=20) as loader:
+        _check_simple_reader(loader, synthetic_dataset.data, BATCHABLE_FIELDS - {TestSchema.sensor_name})
+
+
+@pytest.mark.parametrize('reader_factory', ALL_READER_FLAVOR_FACTORIES)
```
Comment (selitvin) [inline petastorm/tests/test_pytorch_dataloader.py]: Does this test checks something different comparing to `test_simple_read_with_small_shuffle_size`?

**Comment by selitvin** [comment]
Superseded by #382
